Imports relationships per type

tinwelint opened the pull request:
for better scalability for data sets where most nodes are dense and
containing lots of relationship types.

Problem:
Keeping track of and updating heads of relationship chains in the node -> relationship cache
in scenarios where most nodes are dense and there are many relationship types scales poorly
w/ regards to cache memory usage also as a side-effect time required to update this cache.
Previously all relationship groups would need to be present in the cache and so adding more
relationships of different types to dense nodes would create more and more of these groups
in the cache, linking to each using a pointer into the simple array backing the cache.
Both memory usage would steadily increase as more types were added and finding group
for specific type would be more and more expensive due to the linear and random access
pointer chasing, resulting in poor performance or inability to import at all.

Solution:
The approach is instead to do multiple iterations where each iteration imports relationships
of a single type. For each iteration Node --> Relationship and Relationship --> Relationship
stages _for dense nodes only_ are run so that the cache can be reused to hold relationship chain
heads of the next type in the next iteration. All relationships will be imported this way
and then finally there will be one Node --> Relationship and Relationship --> Relationship stage

tinwelint committed:
Imports relationships per type

for better scalability for data sets where most nodes are dense and
containing lots of relationship types.

Problem:
Keeping track of and updating heads of relationship chains in the node -> relationship cache
in scenarios where most nodes are dense and there are many relationship types scales poorly
w/ regards to cache memory usage also as a side-effect time required to update this cache.
Previously all relationship groups would need to be present in the cache and so adding more
relationships of different types to dense nodes would create more and more of these groups
in the cache, linking to each using a pointer into the simple array backing the cache.
Both memory usage would steadily increase as more types were added and finding group
for specific type would be more and more expensive due to the linear and random access
pointer chasing, resulting in poor performance or inability to import at all.

Solution:
The approach is instead to do multiple iterations where each iteration imports relationships
of a single type. For each iteration Node --> Relationship and Relationship --> Relationship
stages _for dense nodes only_ are run so that the cache can be reused to hold relationship chain
heads of the next type in the next iteration. All relationships will be imported this way
and then finally there will be one Node --> Relationship and Relationship --> Relationship stage
linking all sparse relationship chains together.

tinwelint committed:
Pulls out deletion of duplicate nodes into own stage

as to be able to simplify the code for reading records as input steps in
some stages during import.

tinwelint committed:
Individal id spaces for different token types in InputCache

as well as support for Integer as token key. Previously all token types
shared a single id space (2^16 bits), which would theoretically run out in
some scenarios. This is now more correct and will handle all types of
input.

tinwelint committed:
Properly shuts down log service in StoreMigration

tinwelint committed:
Removed the concept of specific ids

because it clashes conceptually with the per-type import.
This means that migration, when store files actually needs to be
migrated will break any legacy relationship indexes. This will have
to be communicated with a warning during migration and in the manual
optimally.

tinwelint committed:
Inserts minority types with batch insertion style

i.e. relationship types that have less relationships than whatever
the batch size for the importer is. This is because there's a certain
amount of overhead executing the stages for a relationship type and
if there are few relationships of a type then it's not worth it and
inserting using batch-insert (i.e. random access style) is more
efficient.

chrisvest commented:
@tinwelint There is a test failure in `upgrade.StoreMigratorFrom20IT.shouldMigrate(StoreMigratorFrom20IT.java:139)`. This test should also be improved to say _why_ it is failing; e.g. what inconsistencies it has found. That would be a nice addition.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipTypeCheckerStep.java:
```diff
@@ -0,0 +1,142 @@
+        {
+            Object type = relationship.typeAsObject();
+            AtomicLong count = allTypes.get( type );
+            // Check w/o synchronized, it's fine
```
I don't think this is fine. If the HashMap grows, it will re-hash its entries, and this will reverse the linked lists in its buckets. If this coincides with a racy read, the read can observe a loop in the bucket structure. If those values happen to be cached in registers or on the stack, and the value it is looking for is not among the entries that form the loop, the reading thread can get stuck iterating this loop forever.

chrisvest replied:
If the `get` effectively performed a volatile read, and the `put` effectively performed a volatile write, then there would be no problem with this kind of double-checked locking.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -19,72 +19,137 @@
     private static final long EMPTY = -1;
-    private static final byte[] DEFAULT_VALUE = new byte[10];
     private static final long MAX_RELATIONSHIP_ID = (1L << 48/*6B*/) - 2/*reserving -1 as legal default value*/;
+    static final int MAX_COUNT = (1 << 30/*2 change bits*/) - 2/*reserving -1 as legal default value*/;
```
So imported nodes can at most have 1 billion relationships?

tinwelint replied:
Yes, although 2.3 could only have 2^24. We can change it later if it becomes a problem.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -19,72 +19,137 @@
+        assertValidCount( index, count );
+        array.setInt( index, offset, count );
+        return count;
+    }
```
This does not need to be thread safe?

tinwelint replied:
It does not. There are external factors making it so that there's only a single thread at a time changing one particular index in that array. See for example `CalculateDenseNodesStage` and its steps, particularly `CalculateDenseNodePrepareStep`

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -19,72 +19,137 @@
+ *   [NEXT,OUT_ID,OUT_DEGREE,IN_ID,IN_DEGREE,LOOP_ID,LOOP_DEGREE]
+ * - DEGREE < THRESHOLD: last seen relationship id for this node
  */
 public class NodeRelationshipCache implements MemoryStatsVisitor.Visitable
```
The thread-safety of this class needs to be documented. Some parts appear to account for multi-threaded access, while other parts don't. This is confusing.

tinwelint replied:
This documentation could be improved, yes

tinwelint committed:
Uses ConcurrentHashMap in RelationshipTypeCheckerStep

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ReadNodeRecordsByCacheStep.java:
```diff
@@ -0,0 +1,110 @@
+                endOfUpstream();
+            }
+        }.start();
+        return 0;
```
How many threads will this start? Can you use a pool?

tinwelint replied:
Only 1

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -117,28 +193,89 @@ public long getAndPutRelationship( long nodeId, int type, Direction direction, l
+
+    private boolean markAsChanged( ByteArray array, long nodeId, long mask )
+    {
+        long bits = array.getInt( nodeId, SPARSE_COUNT_OFFSET ) & 0xFFFFFFFF;
```
This is wrong. You need to AND with `0xFFFFFFFFL` instead of `0xFFFFFFFF`.

Proof:

```
int x = 0xF0_00_00_00;
long X = x & 0xFFFFFFFF;
long Y = x & 0xFFFFFFFFL;
System.out.println( "X = " + X );
System.out.println( "Y = " + Y );
```

Prints

```
X = -268435456
Y = 4026531840
```

Intellij warns about this. Maybe you can make Eclipse do the same.

tinwelint replied:
You're absolutely correct in principal, if that value were to be interpreted as... a value. Now we're only interested in the (32) bits. In this case it's actually enough to do:

```
int bits = array.getInt( nodeId, SPARSE_COUNT_OFFSET );
```

chrisvest replied:
That would mean that the `(bits & mask) != 0` expression below would up-cast and sign-extend the `bits`, which might not be the bit pattern that you are expecting with the `mask`.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -117,28 +193,89 @@ public long getAndPutRelationship( long nodeId, int type, Direction direction, l
+
+    private static boolean nodeIsChanged( ByteArray array, long nodeId, long mask )
+    {
+        long bits = array.getInt( nodeId, SPARSE_COUNT_OFFSET ) & 0xFFFFFFFF;
```
Same problem here.

tinwelint replied:
As mentioned above... I'll change it to be simpler.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -155,91 +292,101 @@ private static long all48Bits( long raw )
-            }
-        }
-        relGroupCache.clearRelationships();
+        oneMeansChanged = forward;
```
I can't make sense of the name of this field.

tinwelint replied:
Mmh, changed it to `forward` and added comment to that field.

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -117,28 +193,89 @@ public long getAndPutRelationship( long nodeId, int type, Direction direction, l
+                synchronized ( chunkChangedArray )
+                {
+                    chunkChangedArray[chunk] |= mask;
+                }
```
These writes don't synchronise with the reads, because the array reads are not volatile. Not a problem in this method I guess, since the writes are idempotent, and the synchronisation will prevent work tearing. However, do you ensure that all calls to `getAndPutRelationship` happens before the reads in `visitChangedNodes`?

tinwelint replied:
You're right, they are idempotent and `visitChangedNodes` happens after all calls to `getAndPutRelationship`. Documentation...

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -155,91 +292,101 @@ private static long all48Bits( long raw )
+        long visit( long nodeId, long next, long out, long in, long loop );
     }
 
     public static final GroupVisitor NO_GROUP_VISITOR = new GroupVisitor()
```
This can be a lambda.

tinwelint commented:
The test has an inconsistency, looking like the highest node haven't really been updated properly. Perhaps it's an issue with off-by-one high id somewhere.

tinwelint committed:
Added some javadocs to NodeRelationshipCache

tinwelint committed:
Fixes import issue where all relationships where of minority

tinwelint commented:
@chrisvest Addressed comments and found the test issue https://github.com/neo4j/neo4j/pull/7026/commits/75308b60d02ff8cddbb42e484e8c892c333c1174

chrisvest commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -117,28 +205,89 @@ public long getAndPutRelationship( long nodeId, int type, Direction direction, l
+    {
+        int bits = array.getInt( nodeId, SPARSE_COUNT_OFFSET );
+        boolean changeBitIsSet = (bits & mask) != 0;
+        boolean changeBitWasFlipped = changeBitIsSet != forward;
```
Brining this comment out of the commit and into the PR:

You said you are expecting to use all 32 bits of `bits`. During the `(bits & mask)` expression, the `bits` will be up-casted to a `long` because `mask` is `long`, and this will move the sign bit. This might not fit the pattern that you are expecting with `mask`.

tinwelint replied:
Maybe although the tests which checks this behaviour all work. Would making the mask `int` instead make things more straight forward?

chrisvest commented:
@tinwelint I've looked through the code. Only one comment left unaddressed now.

tinwelint committed:
Creates minority relationships with correct ids

following ids of the other relationships, not using the relationship store
since its high id at this point is set to the number of relationships to
create as a whole. Also doesn't call nextId for relationships which has
missing nodes, i.e. bad relationships.

tinwelint committed:
Doesn't reuse original store's id files after migration

since entities may now be moved to different ids.

tinwelint committed:
Clarifies node->rel mask by making it int

chrisvest commented:
Looks good!
